### PR TITLE
IAM 726

### PR DESCRIPTION
**Pull request by shipperizer**
IAM-726: use worker pool for background calls to openFGA

- **feat: add worker pool implementation**
- **feat: hook up worker pool for groups and roles API**

```

shipperizer in ~/shipperizer/jobless-needle on main λ hey -c 50 -n 1000 -H X-Authorization:c2hpcHBlcml6ZXI= http://localhost:8000/api/v0/roles/administrator/entitlements


Summary:
  Total:	20.3209 secs
  Slowest:	0.1557 secs
  Fastest:	0.0025 secs
  Average:	0.0174 secs
  Requests/sec:	49.2105
  
  Total data:	295516 bytes
  Size/request:	303 bytes

Response time histogram:
  0.002 [1]	|
  0.018 [838]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.033 [75]	|■■■■
  0.048 [23]	|■
  0.064 [2]	|
  0.079 [0]	|
  0.094 [0]	|
  0.110 [3]	|
  0.125 [4]	|
  0.140 [9]	|
  0.156 [20]	|■


Latency distribution:
  10% in 0.0067 secs
  25% in 0.0092 secs
  50% in 0.0117 secs
  75% in 0.0150 secs
  90% in 0.0242 secs
  95% in 0.0389 secs
  99% in 0.1538 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0001 secs, 0.0025 secs, 0.1557 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0020 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0019 secs
  resp wait:	0.0173 secs, 0.0024 secs, 0.1537 secs
  resp read:	0.0000 secs, 0.0000 secs, 0.0014 secs

Status code distribution:
  [200]	971 responses
  [500]	4 responses

Error distribution:
  [25]	Get "http://localhost:8000/api/v0/roles/administrator/entitlements": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

shipperizer in ~/shipperizer/jobless-needle on main λ hey -c 50 -n 1000 -H X-Authorization:c2hpcHBlcml6ZXI= http://localhost:8000/api/v0/groups/c-level/entitlements


Summary:
  Total:	20.3215 secs
  Slowest:	0.0902 secs
  Fastest:	0.0022 secs
  Average:	0.0160 secs
  Requests/sec:	49.2090
  
  Total data:	175223 bytes
  Size/request:	179 bytes

Response time histogram:
  0.002 [1]	|
  0.011 [355]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.020 [485]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.029 [61]	|■■■■■
  0.037 [21]	|■■
  0.046 [7]	|■
  0.055 [6]	|
  0.064 [0]	|
  0.073 [11]	|■
  0.081 [14]	|■
  0.090 [13]	|■


Latency distribution:
  10% in 0.0079 secs
  25% in 0.0098 secs
  50% in 0.0124 secs
  75% in 0.0159 secs
  90% in 0.0228 secs
  95% in 0.0395 secs
  99% in 0.0845 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0001 secs, 0.0022 secs, 0.0902 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0035 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0044 secs
  resp wait:	0.0158 secs, 0.0022 secs, 0.0888 secs
  resp read:	0.0001 secs, 0.0000 secs, 0.0058 secs

Status code distribution:
  [200]	973 responses
  [500]	1 responses

Error distribution:
  [26]	Get "http://localhost:8000/api/v0/groups/c-level/entitlements": context deadline exceeded (Client.Timeout exceeded while awaiting headers)


```

## TODO

- [ ] add tests on the pool package
- [ ] validate all tests are still passing
- [ ] create issue to address shared worker pool across APIs

### CODE REVIEW

File: cmd/serve.go [hunk 1]
```diff
@@ -119,7 +120,9 @@ func serve() {
 		}
 	}
 
-	router := web.NewRouter(idpConfig, schemasConfig, rulesConfig, extCfg, web.NewO11yConfig(tracer, monitor, logger))
+	wpool := pool.NewWorkerPool(specs.WorkersTotal, tracer, monitor, logger)
+
+	router := web.NewRouter(idpConfig, schemasConfig, rulesConfig, extCfg, web.NewO11yConfig(tracer, monitor, logger), wpool)
```
Comment (shipperizer) [inline cmd/serve.go:125]:
```suggestion
	router := web.NewRouter(idpConfig, schemasConfig, rulesConfig, extCfg,wpool,  web.NewO11yConfig(tracer, monitor, logger))
```

leave o11y config as the last one in any signature (implicit convention)

File: internal/config/specs.go [hunk 2]
```diff
@@ -38,4 +38,6 @@ type EnvSpec struct {
 	ModelId   string `envconfig:"openfga_authorization_model_id" default:""`
 
 	AuthorizationEnabled bool `envconfig:"authorization_enabled" default:"false"`
+
+	WorkersTotal int `envconfig:"workers_total" default:"25"`
```
Comment (shipperizer) [inline internal/config/specs.go:42]:
```suggestion
	WorkersTotal int `envconfig:"workers_total" default:"150"`
```

boost defautl, see if u can change variable name to something more specific

File: internal/pool/pool.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	jobs chan *job
+
+	shutdownCtx  context.Context
+	shutdownFunc context.CancelCauseFunc
+
+	wg sync.WaitGroup
+
+	Tracer  tracing.TracingInterface
+	Monitor monitoring.MonitorInterface
+	Logger  logging.LoggerInterface
```
Comment (shipperizer) [inline internal/pool/pool.go:34]:
```suggestion
	tracer  tracing.TracingInterface
        monitor monitoring.MonitorInterface
	logger  logging.LoggerInterface
```

keep them private

File: cmd/serve.go [hunk 4]
```diff
@@ -119,7 +120,9 @@ func serve() {
 		}
 	}
 
-	router := web.NewRouter(idpConfig, schemasConfig, rulesConfig, extCfg, web.NewO11yConfig(tracer, monitor, logger))
+	wpool := pool.NewWorkerPool(specs.WorkersTotal, tracer, monitor, logger)
```
Comment (shipperizer) [inline cmd/serve.go:123]:
```suggestion
	wpool := pool.NewWorkerPool(specs.WorkersTotal, tracer, monitor, logger)
	defer wpool.Stop()
```

File: internal/pool/pool.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	ret := make([]Result[T], 0)
+
+	for i := 0; i < n; i++ {
+		r := <-results
+		result := NewResult[T](r.key, r.Value.(T))
+		ret = append(ret, *result)
+	}
+
+	return ret
+}
```
Comment (shipperizer) [inline internal/pool/pool.go:127]:
```suggestion
```

File: pkg/web/router.go [hunk 6]
```diff
@@ -25,7 +26,7 @@ import (
 	"github.com/canonical/identity-platform-admin-ui/pkg/status"
 )
 
-func NewRouter(idpConfig *idp.Config, schemasConfig *schemas.Config, rulesConfig *rules.Config, externalConfig ExternalClientsConfigInterface, ollyConfig O11yConfigInterface) http.Handler {
+func NewRouter(idpConfig *idp.Config, schemasConfig *schemas.Config, rulesConfig *rules.Config, externalConfig ExternalClientsConfigInterface, ollyConfig O11yConfigInterface, wpool pool.WorkerPoolInterface) http.Handler {
```
Comment (shipperizer) [inline pkg/web/router.go:29]:
```suggestion
func NewRouter(idpConfig *idp.Config, schemasConfig *schemas.Config, rulesConfig *rules.Config, externalConfig ExternalClientsConfigInterface, wpool pool.WorkerPoolInterface, ollyConfig O11yConfigInterface) http.Handler {
```

same as above

**Comment by shipperizer** [comment]
:+1

**Review by BarcoMasile — APPROVED** [review]

**Pull request merged**